lib/idtree.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * id allocating tree
+ *
+ * Copyright 2018 Phoenix Systems
+ * Author: Jan Sikorski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "lib.h"
+
+
+static int lib_idtreeCmp(rbnode_t *n1, rbnode_t *n2)
+{
+	idnode_t *i1 = lib_treeof(idnode_t, linkage, n1);
+	idnode_t *i2 = lib_treeof(idnode_t, linkage, n2);
+
+	if (i1->id > i2->id) {
+		return 1;
+	}
+	else if (i2->id > i1->id) {
+		return -1;
+	}
+	else {
+		return 0;
+	}
+}
+
+
+static int lib_idtreeGapcmp(rbnode_t *n1, rbnode_t *n2)
+{
+	idnode_t *r1 = lib_treeof(idnode_t, linkage, n1);
+	idnode_t *r2 = lib_treeof(idnode_t, linkage, n2);
+	rbnode_t *child = NULL;
+	int ret = 1;
+
+	if (r1->lmaxgap > 0 && r1->rmaxgap > 0) {
+		if (r2->id > r1->id) {
```

Comment:
r2-> id is always zero so there's no need for this branch thus this whole if else can be simplified to 
```c
if (r1->lmaxgap > 0) {
		child = n1->left;
		ret = 1;
	}
	else if (r1->rmaxgap > 0) {
		child = n1->right;
		ret = -1;
	}
```

Reply:
Why do you think it's always zero? It's not obligatory to have id == 0 element in the tree:
- allocate id = 0,
- allocate id = 1,
- remove id = 0
and we have "id = 0"-less tree

Reply:
this function is only used in `lib_rbFindEx` where n2 is set to `n` whose id is set to 0. During the execution of `lib_rbFindEx` the second argument is constant thus r2->id is always 0

Reply:
I'd rather have this code more generic than it's necessary right now, so I won't get bitten in the future

Reply:
Ok